### PR TITLE
Dedicated methods to check if file or if directory

### PR DESCRIPTION
**Pull request by dvdalilue**
This PR solves a bug that happens when the `src` parameter contains a null byte.

### CODE REVIEW

**Comment by samlown** [comment]
I don't quite understand this PR. The suggestion is that we're checking for a null byte, yet there is a lot of code for checking the type of file, when it strikes me that the code only needs to check if we're loading a file, or not.

I'm assuming by null byte the issue is that the `src` is just empty. Wouldn't it be easier to just check that?

**Comment by dvdalilue** [comment]
The issue is when `src` is a file itself and not a path, that file can have null bytes in it and `Pathname.new(src)` will result in an error. This PR just avoids that error

**Comment by samlown** [comment]
Okay... so the proposal is to allow a "File Object" to be attached, instead of using a path? The naming here is very confusing.

File: lib/prawn/attachment.rb [hunk 1]
```diff
@@ -54,10 +52,23 @@ def attach(src, options = {})
 
     private
 
-    def data_and_opts(path, src, options)
+    def file?(path)
+      File.file?(path)
+    rescue ArgumentError
```
Comment (samlown) [inline lib/prawn/attachment.rb]: `ArgumentError` should not be handled by internal methods. Perform a data check before the method is called.

File: lib/prawn/attachment.rb [hunk 2]
```diff
@@ -39,11 +39,9 @@ class Error < StandardError; end
     # hidden, then nil is returned.
     #
     def attach(src, options = {})
-      path = Pathname.new(src)
-
-      raise ArgumentError, "Data source can't be a directory" if path.directory?
+      raise ArgumentError, "Data source can't be a directory" if directory?(src)
```
Comment (samlown) [inline lib/prawn/attachment.rb]: This is confusing. We should be testing for what we *want*, not what we *don't want*.

**Comment by dvdalilue** [comment]
It's possible to attach a file, the proposal is to handle an edge case of a file with null bytes

File: lib/prawn/attachment.rb [hunk 3]
```diff
@@ -54,10 +52,23 @@ def attach(src, options = {})
...
+      false
+    end
+
+    def directory?(path)
+      File.directory?(path)
+    rescue ArgumentError
+      false
+    end
+
+    def data_and_opts(src, options)
```
Comment (samlown) [inline lib/prawn/attachment.rb]: I'd try to avoid using methods that return two values in ruby. It's not very idiomatic, and we be clearer to have two simple methods.

**Review by samlown — CHANGES_REQUESTED** [review]

**Comment by samlown** [comment]
> It's possible to attach a file, the proposal is to handle an edge case of a file with null bytes

So, an empty file? Isn't it easier to just test for an empty file?

File: lib/prawn/attachment.rb (hunk 3 above)
Comment (dvdalilue) [reply lib/prawn/attachment.rb]: Ok, now there are two methods for both variables

File: lib/prawn/attachment.rb (hunk 2 above)
Comment (dvdalilue) [reply lib/prawn/attachment.rb]: I'm not entirely sure about this line, should we remove this check? Checking for `file?` is not the same

Comment (samlown) [reply lib/prawn/attachment.rb]: You need to think of this more as "story telling". The basic objective with this entry point is to check if the use has provided a path to a file, or an IO object. Anything else is invalid. So start there:

```suggestion
      raise ArgumentError, "data source must be file path or io object" unless file?(src)
```

File: lib/prawn/attachment.rb [hunk 4]
```diff
@@ -54,16 +53,34 @@ def attach(src, options = {})
 
     private
 
-    def data_and_opts(path, src, options)
-      opts = options.dup
-      return [src, opts] unless path.file?
+    def file?(path)
+      return false if path.include?("\u0000")
```
Comment (samlown) [inline lib/prawn/attachment.rb]: It's not clear at all why this is important or in what cases this might happen. There are no tests which could be used to replicate the issue either. Either way, surely the `File.file?` call below would check to ensure the file exists or raise an error.

File: lib/prawn/attachment.rb (hunk 2 above)
Comment (dvdalilue) [reply lib/prawn/attachment.rb]: Yeah, but `file?` won't be valid for this. I could leave `if directory?(src)` and change only the message

File: lib/prawn/attachment.rb (hunk 4 above)
Comment (dvdalilue) [reply lib/prawn/attachment.rb]: There are files (like `.pdf`) that have null characters and are non-empty. If the null char is present, it's not a file path. I was thinking about a specific `path:` key argument to distinguish a string from a file path.

File: lib/prawn/attachment.rb (hunk 2 above)
Comment (samlown) [reply lib/prawn/attachment.rb]: Why not? That's my point. It's not **obvious** from the code what this method needs.

**Comment by samlown** [comment]
> > It's possible to attach a file, the proposal is to handle an edge case of a file with null bytes
> 
> So, an empty file? Isn't it easier to just test for an empty file?

I'm just re-reading these previous comments. It doesn't seem like we're doing anything useful here. If the call to `Pathname.new(src)` results in an error, then the error is already handled. But we can't really continue with this until there are some tests.

**Comment by samlown** [comment]
An alternative solution for all this, would be make the API less flexible but much more concise. If you just set the expectation of the `attach` method to be a File object that response to `read`, job done. The consumer can provide a File, a StringIO, or whatever they fancy. The API could then change to:

```ruby
attach("sample.xml", File.open("/path"), description: "Test Data")
```

**Comment by dvdalilue** [comment]
> An alternative solution for all this, would be make the API less flexible but much more concise. If you just set the expectation of the `attach` method to be a File object that response to `read`, job done. The consumer can provide a File, a StringIO, or whatever they fancy. The API could then change to:
> 
> ```ruby
> attach("sample.xml", File.open("/path"), description: "Test Data")
> ```

Ok, probably it's better to have an easier approach like this, less magic under the hood.

File: lib/prawn/attachment.rb [hunk 5]
```diff
@@ -18,11 +18,11 @@ class Error < StandardError; end
     # a string representation can be embedded.
     #
     # Arguments:
-    # <tt>src</tt>:: path to file, data string, or an object that responds to #to_str
-    # and #length.
+    # <tt>name</tt>:: name of attached file.
+    # <tt>data</tt>:: data string, or an object that responds to #to_str and
```
Comment (samlown) [inline lib/prawn/attachment.rb:22]: Is `data` correct? I thought we'd agreed upon including a file handler. Sending data objects around is a lot less efficient than IO instances.

File: lib/prawn/attachment.rb [hunk 6]
```diff
@@ -18,11 +18,11 @@ class Error < StandardError; end
...
+    # #length.
     #
     # Options:
-    # <tt>:name</tt>:: explicit default filename override.
     # <tt>:creation_date</tt>:: date when the file was created.
```
Comment (samlown) [inline lib/prawn/attachment.rb:26]: I appreciate these were already done, but since we're changing the API anyway... we need more appropriate naming here. Ruby has naming like `:created_at` and `:updated_at` for timestamps, which is what these are. Date tends to be reserved for short dates, like `2022-03-04`. I'd suggest renaming these to `:created_at` and `:modified_at`, each of which expects a `Time` object.

File: spec/prawn/attachment_spec.rb [hunk 7]
```diff
@@ -9,7 +9,7 @@
     expect do
       Prawn::Document.generate("hello.pdf") do
```
Comment (samlown) [inline spec/prawn/attachment_spec.rb:10]: Since we're working on this, it wouldn't hurt to have more tests for different IO scenarios or testing for the presence of the date, modification times, etc.

**Review by samlown — CHANGES_REQUESTED** [review]
Since the API is changing, we should do more to be consistent.

**Comment by samlown** [comment]
This is now fixed by #3
